homeassistant/components/recorder/purge.py

```diff
@@ -151,17 +374,25 @@ def _select_unused_attributes_ids(
         # > explain select min(attributes_id) from states where attributes_id = 136723;
         # ...Select tables optimized away
         #
-        id_query = session.query(column("id")).from_statement(
-            union(
-                *[
-                    select(func.min(States.attributes_id).label("id")).where(
-                        States.attributes_id == attributes_id
-                    )
-                    for attributes_id in attributes_ids
-                ]
-            )
-        )
-    to_remove = attributes_ids - {state[0] for state in id_query.all()}
+        # We used to generate a query based on how many attribute_ids to find but
+        # that meant sqlalchemy Transparent SQL Compilation Caching was working against
+        # us by cached up to MAX_ROWS_TO_PURGE different statements which could be
```

Comment:
Not familiar with SQL Alchemy, but would it be possible to execute a query without it being cached ? We don't purge that often, so having slightly more expensive query generation is fine.

Reply:
My first thought was to disable the cache.  

Unfortunately the query generates so many ORM objects (also the same reason the cache gets so big), generating it every time was more expensive run-time wise than all the other python code run in a purge cycle combined.

Sqlalchemy clearly isn't optimized for this use case.

I also tried caching a query with many bindparms but the code that generates the actual query from the cache clones the cached version and visits every part to fill in the bind values so that turned out to be even more expensive.

The only way I found that resulted in a single cache key (the lambda) was how it's currently implemented.